zebra-chain/src/orchard/sinsemilla.rs

```diff
@@ -0,0 +1,134 @@
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash
+// XXX: should j be strictly limited to k=10 bits?
+#[allow(non_snake_case)]
+fn S(j: [u8; 2]) -> pallas::Point {
+    pallas_group_hash(b"z.cash:SinsemillaS", &j)
+}
+
+/// "...an algebraic hash function with collision resistance (for fixed input
+/// length) derived from assumed hardness of the Discrete Logarithm Problem on
+/// the Pallas curve."
+///
+/// SinsemillaHash is used in the definitions of Sinsemilla commitments and of
+/// the Sinsemilla hash for the Orchard incremental Merkle tree (§ 5.4.1.3
+/// ‘MerkleCRH^Orchard Hash Function’).
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+// XXX: M is a max of k*c = 2530 bits
+pub fn sinsemilla_hash_to_point(D: &[u8], M: &BitVec<Lsb0, u8>) -> pallas::Point {
+    let k = 10;
+    let c = 253;
+
+    assert!(M.len() <= k * c);
+
+    let mut acc = Q(D);
+
+    // Split M into n segments of k bits, where k = 10 and c = 253, padding
+    // the last segment with zeros.
+    //
+    // https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash
+    for chunk in M.chunks(k) {
+        // Pad each chunk with zeros.
+        let mut store = 0u8;
+        let bits = store.bits_mut::<Lsb0>();
+        chunk
+            .iter()
+            .enumerate()
+            .for_each(|(i, bit)| bits.set(i, *bit));
+
+        // An instance of LEBS2IP_k
+        let j = &bits.iter().fold(0u16, |j, &bit| j * 2 + bit as u16);
+
+        acc = acc + acc + S(j.to_le_bytes());
+    }
+
+    acc
+}
+
+/// Sinsemilla Hash Function
+///
+/// "SinsemillaHash is an algebraic hash function with collision resistance (for
+/// fixed input length) derived from assumed hardness of the Discrete Logarithm
+/// Problem. It is designed by Sean Bowe and Daira Hopwood. The motivation for
+/// introducing a new discrete-log-based hash function (rather than using
+/// PedersenHash) is to make efcient use of the lookups available in recent
+/// proof systems including Halo 2."
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+// XXX: M is a max of k*c = 2530 bits, sinsemilla_hash_to_point checks this
+pub fn sinsemilla_hash(D: &[u8], M: &BitVec<Lsb0, u8>) -> pallas::Base {
+    extract_p(sinsemilla_hash_to_point(D, M))
+}
+
+/// Sinsemilla commit
+///
+/// We construct Sinsemilla commitments by hashing to a point with Sinsemilla
+/// hash, and adding a randomized point on the Pallas curve.
+///
+/// SinsemillaCommit_r(D, M) := SinsemillaHashToPoint(D || "-M", M) + [r]GroupHash^P(D || "-r", "")
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
+#[allow(non_snake_case)]
+pub fn sinsemilla_commit(r: pallas::Scalar, D: &[u8], M: &BitVec<Lsb0, u8>) -> pallas::Point {
+    sinsemilla_hash_to_point(&[D, b"-M"].concat(), M)
+        + pallas_group_hash(&[D, b"r"].concat(), b"") * r
```

Comment:
Are there test vectors for `sinsemilla_commit` yet?

<img width="860" alt="Screen Shot 2021-03-18 at 16 11 02" src="https://user-images.githubusercontent.com/8951843/111581205-91112a00-8804-11eb-87a9-4f0922dfed8a.png">

Missing `-` in `-r`:
```suggestion
        + pallas_group_hash(&[D, b"-r"].concat(), b"") * r
```

Reply:
Great catch! No sinsemilla test vectors yet i think...